apps/opentelemetry/src/otel_batch_olp.erl

```diff
@@ -0,0 +1,623 @@
+                                              tables=Tabs,
+                                              scheduled_delay_ms=SendInterval,
+                                              reg_name=RegName,
+                                              otel_signal=Signal}) ->
+    case do_init_exporter(Signal, RegName, ExporterConfig) of
+        %% error should be retried
+        error ->
+            {keep_state_and_data, [{state_timeout, SendInterval, do_init_exporter}]};
+        ignore ->
+            %% no exporter: disable the insertion of new log events and delete the current table
+            clear_table_and_disable(AtomicRef, Tabs),
+            {next_state, idle, Data#data{exporter=ignore}};
+        Exporter ->
+            TimerRef = start_exporting_timer(SendInterval),
+            {next_state, idle, Data#data{exporter=Exporter, exporter_timer=TimerRef}}
+    end;
+init_exporter(_, _, _) ->
+    %% Ignore any other, e.g, external events like force_flush in this state
+    keep_state_and_data.
+
+idle(enter, _OldState, _Data) ->
+    keep_state_and_data;
+idle(info, {timeout, Ref, export_signals}, Data=#data{exporter_timer=Ref}) ->
+    {next_state, exporting, Data};
+idle(cast, force_flush, #data{exporter=Exporter}=Data) when Exporter =/= ignore ->
+    {next_state, exporting, Data};
+idle(EventType, EventContent, Data) ->
+    handle_event_(idle, EventType, EventContent, Data).
+
+exporting(info, {timeout, Ref, export_signals}, #data{exporter_timer=Ref}) ->
+    {keep_state_and_data, [postpone]};
+exporting(enter, _OldState, Data=#data{atomic_ref=AtomicRef,
+                                       tables=Tabs,
+                                       max_queue_size=MaxSize,
+                                       exporting_timeout_ms=ExportingTimeout,
+                                       scheduled_delay_ms=SendInterval}) ->
+    CurrentTab = ?current_tab(AtomicRef),
+    {Data1, Actions} =
+        case ?get_available(AtomicRef, CurrentTab) of
+            %% No events yet, maximum available capacity, nothing to export
+            MaxSize ->
+                %% The other table may contain residual (late) writes not exported
+                %% during the previous run. If current table is not empty, we don't
+                %% need to check the size of the previous (currently disabled) table,
+                %% since we will switch to it after this exporter run.
+                %% However, if current table remains empty for a long time,
+                %% neither export nor table switch will be triggered, and any
+                %% residual late log events in the previous table would be left
+                %% dangling. To avoid such cases, we check other table size
+                %% and export it if it's not empty.
+                maybe_export_other_table(CurrentTab, Data);
+            _ ->
+                RunnerPid = export_signals(CurrentTab, Data),
```

Comment:
Tiny nit: we may move Pid to the state, since it is not used in idle

Reply:
It's a `state_functions` state machine, so state must be an atom, so moving Pid to the state would require re-writing it to `handle_event_function`. I think it's not worth it.